BlueMapCore/src/main/java/de/bluecolored/bluemap/core/world/block/BlockEntity.java

```diff
@@ -0,0 +1,142 @@
+package de.bluecolored.bluemap.core.world.block;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class BlockEntity {
+    private static final List<String> COMMON_KEYS = List.of("id", "x", "y", "z", "keepPacked");
+
+    private final String id;
+    private final int x, y, z;
+    private final boolean keepPacked;
+    private final Map<String, Object> data;
+
+    public BlockEntity(Map<String, Object> rawData) {
+        this.id = getString(rawData, "id", "");
+        this.x = getInt(rawData, "x", 0);
+        this.y = getInt(rawData, "y", 0);
+        this.z = getInt(rawData, "z", 0);
+        this.keepPacked = getBoolean(rawData, "keepPacked", false);
+
+        this.data = rawData.entrySet().stream()
+                .filter(it -> !COMMON_KEYS.contains(it.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    public int getInt(Map<String, Object> tree, String key, int def) {
+        Object obj = get(tree, key, String.valueOf(def));
+
+        try {
+            return Integer.parseInt(String.valueOf(obj));
+        } catch (Exception e) {
+            return def;
+        }
+    }
+
+    public int getInt(String key, int def) {
+        return getInt(data, key, def);
+    }
+
+    public boolean getBoolean(Map<String, Object> tree, String key, boolean def) {
+        Object obj = get(tree, key, String.valueOf(def));
+
+        try {
+            return Boolean.parseBoolean(String.valueOf(obj));
+        } catch (Exception e) {
+            return def;
+        }
+    }
+
+    public boolean getBoolean(String key, boolean def) {
+        return getBoolean(data, key, def);
+    }
+
+    public String getString(Map<String, Object> tree, String key, String def) {
+        Object obj = get(tree, key, String.valueOf(def));
+
+        try {
+            return String.valueOf(obj);
+        } catch (Exception e) {
+            return def;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public Map<String, Object> getSubtree(Map<String, Object> tree, String key) {
+        Object obj = get(tree, key, Map.of());
+
+        if (!(obj instanceof Map)) {
+            return Map.of();
+        }
+
+        return (Map<String, Object>) obj;
+    }
+
+    public Map<String, Object> getSubtree(String key) {
+        return getSubtree(data, key);
+    }
+
+    public Object get(Map<String, Object> tree, String key, Object def) {
+        return tree.getOrDefault(key, def);
+    }
+
+    public Object get(String key, Object def) {
+        return get(data, key, def);
+    }
+
+    public String getString(String key, String def) {
+        return getString(data, key, def);
+    }
```

Comment:
This doesn't feel right..
A lot of those methods could be `static`, which then begs the question why we need them in the first place, they look like a lenient but expensive way of casting any "object" into the value-type we need.

I understand the issue that you try to solve here: each tile-entity could have a different data-structure and thus can't be represented by one object-type with a fixed set of properties.

My approach for this would be to create a tree of fixed types anyways (e.g. `SignBlockEntity extends BlockEntity`) with well defined properties, and then write a custom `TypeDeserializer<BlockEntity>` which first reads the whole data-tree (i don't think this can be avoided) but then (based on the `id`) creates the correct sub-type of `BlockEntity`.

Also i think instead of trying to convert between value-types (e.g. `Integer.parseInt(String.valueOf(obj))`) we should just cast based on the type that we expect (`(int) obj`) and deal with the cast-exception and treat it as invalid data-format.

Reply:
Done